### ISSUE TITLE
VncViewer closes when I try to connect again

### ISSUE THREAD

**Issue by drigoskalwalker** [ultravnc/UltraVNC#218]
I'm using wayvnc as vnc server in my Linux machine.
I'm using ultraVNC Viewer on my Windows 11 machine.

I'm able to connect succefull and everything works.
But if I close the connection (closing the window) and try to connect again
the viewer shows up a black screen and immediately closes.

When I use other viewers like RealVNC Viewer
I'm able to reconnect as many times I want

Using ultraVNC Viewer I have to restart my vnc server everytime I close the connection and need to connect again.

**Comment by RudiDeVos** [comment]
Version viewer, version wayvnc server ?
Did this also happen with a previous version of wayvnc ?
After you disconnect a realVNC viewer, can you connect with a UltraVNC viewer ?
After you disconnect a UltraVNC viewer can you connect with a realvnc viewer ?

**Comment by drigoskalwalker** [comment]
Hi,
I tested WayVNC 0.8.0 and the master branch, same behavior.
I tested UltraVNC 1.5.0.0, 1.4.3.6 and 1.3.8.1
TigerVNC Viewer has the same behavior.

- Did this also happen with a previous version of wayvnc ? I really don't know, I don't see advantages to go back in the server.

- After you disconnect a realVNC viewer, can you connect with a UltraVNC viewer ? - NO.

- After you disconnect a UltraVNC viewer can you connect with a realvnc viewer ? - YES.

**Comment by RudiDeVos** [comment]
Looks to me that wayvnc is in a special status after a viewer disconnect to prevent viewers ( expect realvnc) to connect again.
As multiple flavors viewers has the same issue i'm expecting that they don't follow 100% rfb or use some subdialect that isn't backward compatible.  I'm unable to debug what happen on the server site, making it hard todo the required changes.

If wayvnc can provide info on what need to be change we will implement it.

Please post this on https://github.com/any1/wayvnc/issues , as multiple viewers has the same issue.

**Comment by any1** [comment]
Just waiting on feedback: https://github.com/any1/wayvnc/issues/335

I've never been able to reproduce this myself.

**Comment by any1** [comment]
I borrowed my girlfriend's computer and installed ultravnc on it. It turns out that it works fine with wayvnc 0.8.0, but it keeps disconnecting if I run wayvnc from the master branch. Other clients that I've tested work fine.

Is there some way to see why it's disconnecting? I tried running it from "Command Prompt" with the `-loglevel 10` option, but it doesn't show me anything.

**Comment by any1** [comment]
> If you can run a public wayvnc server somewhere I can try debugging UltraVNC viewer problems directly in Visual Studio.

I could set up such a thing, but first I think I'll fix the crash that you ran into. It shouldn't crash if the system doesn't have a render node.

Did you run it with the `-g/--gpu` flag? If so, try removing it. It shouldn't even try to open the render node unless the user sets the gpu flag.

Thanks!

**Comment by any1** [comment]
Yeah, it looks like it tries to open a render node regardless of whether gpu mode is enabled or not. If you run `meson setup build -Dscreencopy-dmabuf=disabled`, it will remove this code path entirely.

**Comment by wqweto** [comment]
UltraVNC log from crashing session
```
Started and Winsock (v 2) initialised
bufsize expanded to 4352
Registered connection with app
Saving to C:\Work\Temp\ultravnc\vncviewer\Debug\options.vnc
DPI 120
Connected to 172.17.17.159 port 5900
DSMPlugin not enabled
RFB server supports protocol version 3.8
Connected to RFB server, using protocol version 3.8
TLSSession: Using TLS_AES_256_GCM_SHA384 (0x1302) ciphersuite
TLSSession: TLS1_3_CLIENT using AES_256 cipher with SHA_384 hash and 0x0 key-exchange
VNC authentication succeeded
Read a 6-byte string
Desktop name "WayVNC "
Geometry 954 x 741 depth 24
Reset keyboard for first use
Flush dead key gives: 1 character(s): 0x0020 ( )
Screen work area is 2560 x 1380
Clipboard changed
Ignore Clipboard while initializing!
No next window in chain; WM_DRAWCLIPBOARD will not be passed
SetClipboardViewer to 0x000a1de0; next is 0x00000000. Last error 0x00000000Cache: Cache buffer bitmap creation
Screen work area is 2560 x 1380
DPI changed, SizeWindow
Screen work area is 2560 x 1380
Update-processing thread started
Screen work area is 2560 x 1380
Screen work area is 2560 x 1380
Read remote clipboard change
Read remote extended clipboard change
Read remote clipboard change
Read a 16-byte string
Unknown message type xb9
Clipboard changed
```

**Comment by any1** [comment]
I actually had it failing even without any kind of encryption. Does it behave the same way if you run it without TLS?

**Comment by any1** [comment]
I think it's likelier to have something to do with the clipboard. For example, if wayvnc sends the wrong size for the clipboard data, the stream will go out of sync.

**Comment by wqweto** [comment]
> I actually had it failing even without any kind of encryption. Does it behave the same way if you run it without TLS?

Yes, disabling `rfbVeNCypt` (i.e. using `rfbRSAAES`) fails the same unknown message 185.

Here is a dump of the decrypted stream:
```
TLSPlugin: decPlain: 00-00-00-00
Handling msgType: 0x00
TLSPlugin: decPlain: 03-ba-02-e5-ff-ff-fe-cc
TLSPlugin: decPlain: 00-00-00-00-03-ba-02-e5-00-00-00-00
TLSPlugin: decPlain: 78-01-ec-bd-09-b8-65-57-59-e6-7f-cf-b9-35-85-4c
TLSPlugin: decPlain: 55-75-6f-2a-13-21-65-55-86-4a-2a-95-00-49-c8-00
TLSPlugin: decPlain: 01-fb-b1-55-7c-da-b6-7d-ba-bb-42-1a-6c-04-9a-d6
TLSPlugin: decPlain: 56-6c-1c-9a-a6-2a-89-c0-3d-b7-98-94-b9-05-44-41
TLSPlugin: decPlain: 10-04-64-10-64-e8-3f-2a-06-44-8c-24-92-74-18-c2
TLSPlugin: decPlain: 24-60-4c-22-12-02-41-42-48-02-04-f0-fe-7f-ef-fb
TLSPlugin: decPlain: ad-b5-f6-da-fb-9c-73-eb-56-a5-6e-52-55-9c-aa-7d
TLSPlugin: decPlain: f7-59-eb-5b-df-fa-86-f7-fb-d6-da-6b-0f-67-9f-a9
TLSPlugin: decPlain: df-db-f9-d3-07-fa-b6-63-fb-96-03-77-03-fc-03-d7
TLSPlugin: decPlain: 78-2c-9f-d8-7f-ff-86-6f-82-ff-04-ff-7b-83-c0-f2
TLSPlugin: decPlain: e5-cf-b3-2f-d9-72-cd-1b-1e-7f-e7-97-3f-7a-cf-3d
TLSPlugin: decPlain: 77-7f-ea-a6-db-e6-e7-5e-b2-63-fb-e9-f7-c6-d4-91
TLSPlugin: decPlain: 7d-97-c9-fe-f9-8b-b7-dc-fa-fe-5f-fc-c0-27-bf-f4
TLSPlugin: decPlain: ec-ff-fd-d4-67-3e-f6-21-1f-7e-fe-4f-5c-ff-d5-4f
TLSPlugin: decPlain: ec-fc-ef-ff-7b-a4-0d-f7-86-d8-b1-ff-f6-2f-bf-f6
TLSPlugin: decPlain: f2-5d-6f-29-02-77-6e-df-72-c3-eb-7e-ec-6d-d7-de
TLSPlugin: decPlain: b1-e3-e2-0b-5e-f7-3f-b6-de-f6-89-57-7f-f7-9e-3b
TLSPlugin: decPlain: 3e-fa-f9-af-de-f2-57-af-7c-eb-b5-77-bd-e2-92-73
TLSPlugin: decPlain: c7-f1-ff-cd-33-1f-f5-a1-1b-ff-61-c7-cf-5d-12-a2
...
<< snip ~16KB of data >>
...
TLSPlugin: decPlain: 65-e3-a0-5a-6d-c5-88-5d-32-ad-c0-d1-a4-b7-48-ad
TLSPlugin: decPlain: 03-b0-18-9b-45-89-19-cd-e2-d4-a8-0f-bf-0c-5a-e2
TLSPlugin: decPlain: a4-a9-5d-28-7e-a2-0b-e2-eb-c6-ae-f8-36-55-cb-3d
TLSPlugin: decPlain: 16-85-32-d3-b9-55-43-8a-03-b2-23-9a-92-83-69-9c
TLSPlugin: decPlain: a5-29-bf-0b-63-76-51-46-39-34-1e-6b-5a-9f-28-4a
Handling msgType: 0x00
Handling msgType: 0x03
Read remote clipboard change
Read a 16-byte string
Handling msgType: 0xb9
Clipboard changed
Unknown message type xb9
We changed it - ignore!
```

**Comment by any1** [comment]
You can see the contents of the clipboard on the "remote" server by running `wl-paste` and/or `wl-paste -p`. Are there actually 16 bytes in there?

**Comment by wqweto** [comment]
Btw, `ff-ff-fe-cc` above is `rfbEncodingExtDesktopSize` so it correctly receives server screen dimensions.

**Comment by any1** [comment]
Thanks for your help! I think this might be enough to get me onto the right track.

If you're in the mood for further investigation, you could try disabling some extensions in neatvnc. E.g. you can disable extended clipboard by commenting it out here: https://github.com/any1/neatvnc/blob/master/src/server.c#L642

**Comment by wqweto** [comment]
There is a bug in `rfbEncodingExtDesktopSize` handling in CLientConnection.cpp: [this `break`](https://github.com/ultravnc/UltraVNC/blob/main/vncviewer/ClientConnection.cpp#L5908) should be a `continue`.

The repercussions for wayvnc is that you should not coalesce `rfbFramebufferUpdate`s i.e. don't encode several "screens" updates in `rfbFramebufferUpdate` when first (one) of them is `rfbEncodingExtDesktopSize`.

**Comment by any1** [comment]
Ahh, I see. It looks like those `break`s are happening in other places as well.

It is convenient to coalesce rects because it simplifies bookkeeping with regard to pending requests, but I'll see if I can find a way to do it that's not too convoluted.

**Comment by drigoskalwalker** [comment]
abreucorp.ddns.net:65535
username:vnctest
password: vnctest

Just try to access if you want, without RealVNC Viewer is not possible.

**Comment by any1** [comment]
I found out by connecting to @drigoskalwalker 's server that TigerVNC doesn't like 24 bit buffers. Does UltraVNC have the same problem?

I suspect that the reason why it connects once and then never again is that wayvnc sets a placeholder buffer which will always be 32 bits (IIRC), but this gets wiped as soon as the first picture is captured. Maybe this gives the client a chance to send a message to change the pixel format before the server sends the first 24 bit buffer?

**Comment by drigoskalwalker** [comment]
UltraVNC has the same problem, only the first connection (after the WayVNC server starts) works.
Afther that only RealVNC Viewer works.

**Comment by RudiDeVos** [comment]
vncviewer with patch build
( internal test builds)
https://www.uvnc.eu/download/1600

**Comment by drigoskalwalker** [comment]
> vncviewer with patch build ( internal test builds) https://www.uvnc.eu/download/1600

Thanks for your help but it didn't work.
You can try yourself:
https://github.com/ultravnc/UltraVNC/issues/218#issuecomment-2468488666

**Comment by any1** [comment]
My last comment may have escaped your attention. The original issue still remains. What was solved is in fact a separate issue.

**Comment by drigoskalwalker** [comment]
> My last comment may have escaped your attention. The original issue still remains. What was solved is in fact a separate issue.

Sorry, my mistake

**Comment by wqweto** [comment]
> I found out by connecting to @drigoskalwalker 's server that TigerVNC doesn't like 24 bit buffers. Does UltraVNC have the same problem?

Yes, `ClientConnection::zrleDecode` does not support 24 in `m_myFormat.bitsPerPixel`. 

There is an issue is with `ClientConnection::Createdib` where `CreateDIBSection` fails for 24 bpp in `bi.bmiHeader.biBitCount` but the result is not checked and `memset` is called on `nullptr`.

Overall I don't think the viewer is tested with 24 bpp buffers at all.

**Comment by any1** [comment]
There are two quick and easy solutions for this
 1. Tell the server to send a different format before requesting the first framebuffer
 2. Convert to 32 bpp. The client knows how to do this anyhow since it must be able to decode "cpixels"

**Comment by Neustradamus** [comment]
@RudiDeVos: Can you look about the second bug?
-> Depth / Bits Per Pixel

**Comment by wqweto** [comment]
@drigoskalwalker Can you please run your wayvnc server once again so I can test @any1's suggestion?

**Comment by drigoskalwalker** [comment]
> @drigoskalwalker Can you please run your wayvnc server once again so I can test @any1's suggestion?

You can test now.

**Comment by any1** [comment]
It turns out that rfc6143 specifies that bpp should be 8, 16 or 32, so I guess the server should just report 32 when the buffers are 24 bits.

**Comment by any1** [comment]
See this for more context: https://github.com/TigerVNC/tigervnc/issues/1867

I'll fix it on my end.

**Comment by wqweto** [comment]
@any1 I just patched the viewer not to fall flat on its face when server proposes unsupported bpp which fixes the issue with wayvnc current and probably past versions.

**Comment by any1** [comment]
> @any1 I just patched the viewer not to fall flat on its face when server proposes unsupported bpp which fixes the issue with wayvnc current and probably past versions.

I appreciate it, but do you really want to do that though? The server is in the wrong here and I don't think that we have many NVidia users. Like Pierre points out, the checks are good for detecting protocol errors. Then again, UltraVNC doesn't seem to report those errors to the user.

**Comment by wqweto** [comment]
> Then again, UltraVNC doesn't seem to report those errors to the user.

Yes, next best thing to reporting the protocol violation would be to just work I guess :-))

**Comment by Neustradamus** [comment]
It is not possible to upgrade/convert 24 to 32?

**Comment by wqweto** [comment]
@Neustradamus IMO both fixes are worth having.

Keep in mind that RealVNC viewer was handling 24 bbp from server fine, probably by converging to 32 bpp the way 639e1861333bd8510c40cd7ecc694a9be39bb246 above does.
